x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizer.java

```diff
@@ -481,59 +482,33 @@ private static Limit descendantLimit(UnaryPlan unary) {
         }
     }
 
-    static class DuplicateLimitAfterMvExpand extends OptimizerRules.OptimizerRule<Limit> {
+    private static boolean isTransparentForLimit(LogicalPlan plan) {
+        return plan instanceof Eval || plan instanceof Project || plan instanceof RegexExtract || plan instanceof Enrich;
+    }
 
+    protected static class PushDownMvExpandPastOrderBy extends OptimizerRules.OptimizerRule<OrderBy> {
```

Comment:
> suggestion: a small example as javadoc would immensely help understand how this rule works.

My advice: take a look at the tests in LogicalPlanOptimizerTests and their javadoc, where the optimized plan is visible. Alternatively, `trace` logging on `org.elasticsearch.xpack.esql.optimizer` to see what each rule is doing when.

> Question: how is pushing MvExpand past an OrderBy beneficial?

In the description of the PR I have identified two cases for this type of pushdown. If the semantics of the query remains the same (and they are in those two situations I mentioned), it is beneficial to form a TopN by having sort (OrderBy) closer to a limit. Since every query now has a `limit`, the aim is to have the sort closer to it.

Reply:
Thanks.

I still think it should be obvious what a rule does from either its name or a javadoc; the optimizer tests are not sufficient as they contain complete optimization runs, so there's no fast way to see what a rule does (in isolation). This rule is also sufficiently complex that skimming the code does not give an immediate idea of what to expect in terms of its behavior.

Reply:
Wait, sorry - this is actually one of the simpler rules - I thought I left this comment further down.

Still - pushing down mv_expand past a sort is not an obvious optimization (obvious would be e.g. pushing down filters to reduce number of rows). A short comment explaining the reasoning for why it exists in the first place would be important IMHO.